Updates to Documentation

valentinaschueller opened the pull request:
Close #25 

@uwefladrich Most of these are small changes - but you might want to look at the [updated Logging Policy in the Developer Guide](https://github.com/valentinaschueller/scriptengine-tasks-ecearth/blob/documentation/docs/sphinx/source/tasks/monitoring/developer-guide.rst).

uwefladrich commented:
Some comments about the logging section:

> Every (monitoring) task is responsible for logging its execution.

Okay.

> In the beginning of self.run() (after very few lines of code), a task must call self.log_info() to log that it is active.

Okay.
I would add: "It may preferably include information about it's arguments in the log message, but nothing that would require a lot of code processing the arguments (which could fail)."

> A task should write only one log_info message during execution.

I am not sure about that (even if I might have said this myself). There may be situations when you want to let the user know something that is not a warning or error. Such as logging into a service, starting another process, creating a file (other than the obvious files that a task is expected to open), etc. Nothing that I see right now for the monitoring tasks, but it could happen. So I think the statement is a bit too strict. Of course, info logging should not be noisy! So lets say something to the effect that tasks should be logging once to info unless there is important information that the user needs to know (other than warning, info)?

> General "progression" statements must be log_debug messages.

Okay.

> These should not be used sparsely.

I do not understand this sentence. Do you mean that there should be _a lot_ of debugging messages?

> If a monitoring task has to abort: It should use log_error

Here, the more strict **must** would be appropriate, I think. If you do not log the error, the user will only get (a more general) error from an instance higher up. So basically a good part of the error information is lost.

See below a comment on "abort".

> and throw one of the ScriptEngine Task Exceptions,

Okay.

> unless the problem is unexpected.

Okay, but somewhat unnecessary. Exceptions can occur in basically every statement.

> If a monitoring tasks faces a problem but does not need to abort, it should use log_warning.

Hm yes, but be careful not to identify an error in the task and "abort". If the task experiences an unrecoverable error, then it is supposed to throw a `ScriptEngineTaskException`, But that doesn't necessarily mean "abort". The scriptengine instance may decide to continue with other tasks or do whatever..

valentinaschueller commented:
First, to clarify: Everything I didn't change in this pull request comes from our last discussion on a logging policy (I actually think you looked through this before :wink:)

> I would add: "It may preferably include information about it's arguments in the log message, but nothing that would require a lot of code processing the arguments (which could fail)."

I will add this

> So lets say something to the effect that tasks should be logging once to info unless there is important information that the user needs to know (other than warning, info)?

I will think about how to update this sentence with this in mind.

> I do not understand this sentence. Do you mean that there should be a lot of debugging messages?

Yes. Not in the sense that there should be as many as possible - but that one should not hesitate to use them and that log_debug() messages should give enough progression statements to understand what's going on. I will probably reformulate the sentence to make that clear

> Here, the more strict must would be appropriate, I think. If you do not log the error, the user will only get (a more general) error from an instance higher up. So basically a good part of the error information is lost.

I will update this.

> Okay, but somewhat unnecessary. Exceptions can occur in basically every statement.

You're right, I will remove this part!

> Hm yes, but be careful not to identify an error in the task and "abort". 

I will clarify this!

I'll update this PR tomorrow!

uwefladrich commented:
> First, to clarify: Everything I didn't change in this pull request comes from our last discussion on a logging policy (I actually think you looked through this before wink)

I see. I will probably complain again next time I read it anyway. :grin: 

> [...]
> > I do not understand this sentence. Do you mean that there should be a lot of debugging messages?
> Yes. Not in the sense that there should be as many as possible - but that one should not hesitate to use them and that log_debug() messages should give enough progression statements to understand what's going on. I will probably reformulate the sentence to make that clear

Okay, but keep in mind that there may be _a lot_ of tasks in a script (counting looping, includes, etc), so one can get drowned in debug messages. So don't be too encouraging either. :smiley:

valentinaschueller commented:
I have now worked on the paragraph:

* On further thinking, I have not changed the line "A task should write only one log_info message during execution." I think the situations you described are quite clear cases for exceptions from this rule (and this is also why it already is a "should", not a "must")
* I removed the line "These should not be used sparsely." From the sentence before (`log_debug` messages are general progression statements), it is already clear when to use them and we do not need to encourage excessive use of `log_debug` messages
* To deal with the abort/`log_error`/`log_warning` discussion, I changed the phrasing as follows:

> If a monitoring task experiences an unrecoverable error: It **must** use `log_error` and throw one of the ScriptEngine Task Exceptions. For problems which do not lead to a `ScriptEngineTaskException`, use `log_warning` instead.

I think this makes the intent clear.
